crates/shared/src/sources/uniswap_v3/event_fetching.rs

```diff
@@ -0,0 +1,271 @@
+    Web3,
+};
+use hex_literal::hex;
+
+use anyhow::{Context, Result};
+use contracts::{
+    uniswap_v3_pool::event_data::{Burn, Mint, Swap},
+    UniswapV3Pool,
+};
+use ethcontract::{
+    common::abi::Error, contract::ParseLog, dyns::DynAllEventsBuilder, errors::ExecutionError,
+    Event, RawLog, H160, H256,
+};
+
+const SWAP_TOPIC: [u8; 32] =
+    hex!("c42079f94a6350d7e6235f29174924f928cc2ac818eb64fed8004e115fbcca67");
+const BURN_TOPIC: [u8; 32] =
+    hex!("0c396cd989a39f4459b5fa1aed6a9a8dcdbc45908acfd67e028cd568da98982c");
+const MINT_TOPIC: [u8; 32] =
+    hex!("7a53080ba414158be7ec69b987b5fb7d07dee101fe85488f0853ae16239d0bde");
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum UniswapV3Event {
+    Burn(Burn),
+    Mint(Mint),
+    Swap(Swap),
+}
+
+impl ParseLog for UniswapV3Event {
+    fn parse_log(log: RawLog) -> Result<Self, ExecutionError> {
+        let standard_event: Option<Result<UniswapV3Event, ExecutionError>> =
+            log.topics.get(0).copied().map(|topic| match topic {
+                H256(BURN_TOPIC) => Ok(UniswapV3Event::Burn(
+                    log.clone().decode(
+                        UniswapV3Pool::raw_contract()
+                            .abi
+                            .event("Burn")
+                            .expect("generated event decode"),
+                    )?,
+                )),
+                H256(MINT_TOPIC) => Ok(UniswapV3Event::Mint(
+                    log.clone().decode(
+                        UniswapV3Pool::raw_contract()
+                            .abi
+                            .event("Mint")
+                            .expect("generated event decode"),
+                    )?,
+                )),
+                H256(SWAP_TOPIC) => Ok(UniswapV3Event::Swap(
+                    log.clone().decode(
+                        UniswapV3Pool::raw_contract()
+                            .abi
+                            .event("Swap")
+                            .expect("generated event decode"),
+                    )?,
+                )),
+                _ => Err(ExecutionError::from(Error::InvalidData)),
+            });
+        if let Some(Ok(data)) = standard_event {
+            return Ok(data);
+        }
+        Err(ExecutionError::from(Error::InvalidData))
+    }
+}
+
+pub struct UniswapV3PoolEventFetcher(pub Web3);
+
+impl EventRetrieving for UniswapV3PoolEventFetcher {
+    type Event = UniswapV3Event;
+    fn get_events(&self) -> DynAllEventsBuilder<Self::Event> {
+        let mut events = DynAllEventsBuilder::new(self.0.clone(), H160::default(), None);
+        let events_signatures = vec![H256(SWAP_TOPIC), H256(BURN_TOPIC), H256(MINT_TOPIC)];
+        events.filter = events
+            .filter
+            .address(vec![])
+            .topic0(events_signatures.into());
+        events
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct RecentEventsCache {
+    /// Block number used as a Key
+    events: BTreeMap<u64, Vec<Event<UniswapV3Event>>>,
```

Comment:
Instead of using just the block as the key it might be nicer to use the block and event index. We already expect the inner events in the block to be in the correct order (by log index) and that way this is encoded in the type and we can get rid of the Vec.

Reply:
You would use `BTreeMap<(u64, usize), Event<UniswapV3Event>>` ?
How would the implementation go for `remove_events_older_than_block` if only block number is provided and log indexes are unknown, meaning we don't know the full keys?

Reply:
Yes, something like `(u64, usize)`. Probably not usize for the log index but similar.

For `remove_events_older_than_block` you can `split_off` at `(block +1, 0)` so that the return value contains all the blocks you want to keep.

Reply:
Implemented.